reference/performance.mdx

```diff
@@ -5,36 +5,288 @@ sidebarTitle: Performance
+
+## Handling mixed types in the same field
+
+### Problem
+
+A single field sometimes stores different data types across events (for instance, strings in some events and integers in others). This is typically a side effect of using “kitchen-sink” ingestion or inconsistent parsing logic in your code.
+
+### Why it matters
+
+- **Reduced compression**: Storing multiple types in the same column (variant column) is less efficient than storing a single type.
+- **Complex queries**: You might need frequent casting or conditional logic in queries (`tostring()` calls, etc.).
+
+### How to fix it
+
+- **Standardize your types at ingest:** If a field is semantically an integer, always send it as an integer.
+- **Use consistent schemas across services:** If multiple applications write to the same dataset, agree on a schema and data types.
+- **Perform corrections at the source:** If you discover your data has been mixed historically, stop ingesting mismatched types. Over time, new blocks will reflect the corrected types even though historical blocks remain mixed.
+
+## Overly wide queries returning more fields than needed
+
+### Problem
+
+By default, Axiom’s query engine projects all fields (`project *`) for each matching event. This can return large amounts of unneeded data, especially in wide datasets with many columns.
+
+### Why it matters
+
+- **High I/O and memory usage**: Unnecessary data is scanned, read, and returned.
+- **Slower queries**: Time is wasted processing fields you never use.
+
+### How to fix it
+
+- **Use `project` or `project-keep`**
+    
+    Specify exactly which fields you need. For example:
+    
+    ```kusto
+    dataset
+    | where status == 500
+    | project timestamp, error_code, user_id
+    ```
+    
+- **Use `project-away` if you only need to exclude a few fields:** If you need 90% of the fields but want to exclude the largest ones, for instance:
+    
+    ```kusto
+    dataset
+    | project-away debug_payload, large_object_field
+    
+    ```
+    
+- **Limit your results**
+    
+    If you only need a sample of events for debugging, use a lower `limit` value (such as 100) instead of the default 1000.
+    
+
+## Regular expressions when simple filters suffice
+
+### Problem
+
+Regular expressions (`matches`, `regex`) can be powerful, but they are also expensive to evaluate, especially on large datasets.
+
+### Why it matters
+
+- **High CPU usage**: Regex filters require complex per-row matching.
+- **Slower queries**: Large swaths of data are scanned with less efficient matching.
+
+### How to fix it
+
+- **Use direct string filters**
+    
+    Instead of:
+    
+    ```kusto
+    dataset
+    | where message matches "[Ff]ailed"
+    ```
+
+    Use:
+    
+    ```kusto
+    dataset
+    | where message contains "failed"
+    ```
+    
+- **Use `search` for substring search:**
+    
+    If you simply need to find “foobar” anywhere in the text, use:
+    
+    ```kusto
+    dataset
+    | search "foobar"
```

Comment:
Search is far more expensive than using a regex on a single field, because it needs to match *all* fields.